File: app/src/main/java/com/kavir/ali/kavir_msg/utility/Utility.java
package com.kavir.ali.kavir_msg.utility;

import android.Manifest;
import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Typeface;
import android.net.Uri;
import android.os.Build;
import android.support.v4.app.ActivityCompat;
import android.telephony.TelephonyManager;
import android.util.Log;
import android.view.Gravity;
import android.view.View;

import com.kavir.ali.kavir_msg.BuildConfig;
import com.kavir.ali.kavir_msg.R;
import com.kavir.ali.kavir_msg.models.SecretCodeModel;
import com.kavir.ali.kavir_msg.server.LoginServer;
import com.tapadoo.alerter.Alerter;
import com.valdesekamdem.library.mdtoast.MDToast;

import java.net.NetworkInterface;
import java.util.Collections;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class Utility {
    public static String getUniqueIMEIId(Context context) {
        try {
            TelephonyManager telephonyManager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
                // TODO: Consider calling
                //    ActivityCompat#requestPermissions
                // here to request the missing permissions, and then overriding
                //   public void onRequestPermissionsResult(int requestCode, String[] permissions,
                //                                          int[] grantResults)
                // to handle the case where the user grants the permission. See the documentation
                // for ActivityCompat#requestPermissions for more details.
                return "";
            }
            String imei = null;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                imei = telephonyManager.getImei();
            } else {
                imei = telephonyManager.getDeviceId();
            }
            Log.e("imei", "=" + imei);
            if (imei != null && !imei.isEmpty()) {
                return imei;
            } else {
                return android.os.Build.SERIAL;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "not_found";
    }



    public static String getMacAddr() {
        try {
            List<NetworkInterface> all = Collections.list(NetworkInterface.getNetworkInterfaces());
            for (NetworkInterface nif : all) {
                if (!nif.getName().equalsIgnoreCase("wlan0")) continue;

                byte[] macBytes = nif.getHardwareAddress();
                if (macBytes == null) {
                    return "";
                }

                StringBuilder res1 = new StringBuilder();
                for (byte b : macBytes) {
                    res1.append(String.format("%02X:",b));
                }

                if (res1.length() > 0) {
                    res1.deleteCharAt(res1.length() - 1);
                }
                return res1.toString().trim().replace(":","");
            }
        } catch (Exception ex) {
        }
        return "02:00:00:00:00:00";
    }


    public static String calSCode(Context context,String[] numberPhone){
        try {
            int n10 = Integer.parseInt(numberPhone[(numberPhone.length)-1]);
            int n9  = Integer.parseInt(numberPhone[(numberPhone.length)-2]);
            String sCode = SaveItem.getItem(context,SaveItem.raw_Scode,"");
            Log.e("sCode:",sCode);
            if (sCode.length()>0){
                int index = n9+n10;
                return sCode.substring(index,index+10);
            }
            return "";
        }catch (Exception e){
            return "";
        }
    }

    public static void getNewVersion(final Activity context){
        Retrofit retrofit = RetrofitClientInstance.getRetrofitInstance();
        LoginServer loginServer = retrofit.create(LoginServer.class);
        loginServer.getSecretCode(String.valueOf(BuildConfig.VERSION_CODE)).enqueue(new Callback<SecretCodeModel>() {
            @Override
            public void onResponse(Call<SecretCodeModel> call, Response<SecretCodeModel> response) {
                if (response.body().getResult().equalsIgnoreCase("success")){
                    checkNewVersion(context,response.body());
                }else {
                    MDToast.makeText(context,response.body().getMessage(),2500,MDToast.TYPE_INFO).show();
                }
            }

            @Override
            public void onFailure(Call<SecretCodeModel> call, Throwable t) {
                Log.e("onFailure:",t.toString());
            }
        });
    }

    public static void getSecretCode(final Activity context){
        Retrofit retrofit = RetrofitClientInstance.getRetrofitInstance();
        LoginServer loginServer = retrofit.create(LoginServer.class);
        loginServer.getSecretCode(String.valueOf(BuildConfig.VERSION_CODE)).enqueue(new Callback<SecretCodeModel>() {
            @Override
            public void onResponse(Call<SecretCodeModel> call, Response<SecretCodeModel> response) {
                if (response.body().getResult().equalsIgnoreCase("success")){
                    String s_raw = response.body().getSecretCode().trim();
                    SaveItem.setItem(context,SaveItem.raw_Scode,s_raw.substring(0,30));
                }else {
                    MDToast.makeText(context,response.body().getMessage(),2500,MDToast.TYPE_INFO).show();
                }
            }

            @Override
            public void onFailure(Call<SecretCodeModel> call, Throwable t) {
                Log.e("onFailure:",t.toString());
            }
        });
    }

    private static void checkNewVersion(Activity context,SecretCodeModel body) {
        if (body.getVersionCode()==null){
            return;
        }
        if (Integer.parseInt(body.getVersionCode())> BuildConfig.VERSION_CODE){
            SaveItem.setItem(context,SaveItem.COME_NEW_VERSION,"1");
            SaveItem.setItem(context,SaveItem.NEW_VERSION_NAME,body.getVersionName());
            SaveItem.setItem(context,SaveItem.NEW_VERSION_URL,body.getUpdateUrl());
        }else{
            SaveItem.setItem(context,SaveItem.COME_NEW_VERSION,"0");
        }
    }

    public static void showAlertNewVersion(final Activity context, String versionName, final String updateUrl) {
        Alerter.create(context)
                .setTitle(context.getString(R.string.new_version))
                .setText(context.getString(R.string.version)+" "+versionName+" "+context.getString(R.string.available))
                .setDuration(10000)
                .setContentGravity(Gravity.CENTER)
                .setTitleTypeface(Typeface.createFromAsset(context.getAssets(), "fonts/Vazir.ttf"))
                .setTextTypeface(Typeface.createFromAsset(context.getAssets(), "fonts/sans.ttf"))
                .setIcon(R.drawable.ic_notifications_black_24dp)
                .setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Uri uri = Uri.parse(updateUrl);
                        Intent goToMarket = new Intent(Intent.ACTION_VIEW, uri);
                        goToMarket.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY |
                                Intent.FLAG_ACTIVITY_NEW_DOCUMENT |
                                Intent.FLAG_ACTIVITY_MULTIPLE_TASK);
                        try {
                            context.startActivity(goToMarket);
                        } catch (ActivityNotFoundException e) {
                            context.startActivity(new Intent(Intent.ACTION_VIEW,
                                    Uri.parse(updateUrl)));
                        }
                    }
                })
                .show();
    }

    public static String calMID(String[] numberPhone){
        try{
            int n10 = Integer.parseInt(numberPhone[(numberPhone.length)-1]);
            int n9  = Integer.parseInt(numberPhone[(numberPhone.length)-2]);
            String[] mac = Utility.getMacAddr().split("");
            return n9+mac[n9+1]+n10+mac[n10+1]+"";


        }catch (Exception e){
            return "";
        }
    }

    public static String calApkId(Context context,String[] numberPhone){
        try {
            String mid = calMID(numberPhone);
            String uid = SaveItem.getItem(context,SaveItem.USER_ID,"");
            int len = uid.length();
            SaveItem.setItem(context,SaveItem.APK_ID,mid+len+uid);
            return mid+len+uid;
        }catch (Exception e){
            return "";
        }
    }





}

File: app/src/main/java/com/kavirelectronic/ali/kavir_info/utility/Utility.kt
package com.kavirelectronic.ali.kavir_info.utility

import android.Manifest
import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Typeface
import android.net.Uri
import android.os.Build
import androidx.core.app.ActivityCompat
import android.telephony.TelephonyManager
import android.util.Log
import android.view.Gravity
import android.view.View
import com.kavirelectronic.ali.kavir_info.BuildConfig
import com.kavirelectronic.ali.kavir_info.R
import com.kavirelectronic.ali.kavir_info.models.SecretCodeModel
import com.kavirelectronic.ali.kavir_info.server.LoginServer
import com.kavirelectronic.ali.kavir_info.utility.RetrofitClientInstance.retrofitInstance
import com.kavirelectronic.ali.kavir_info.utility.SaveItem.getItem
import com.kavirelectronic.ali.kavir_info.utility.SaveItem.setItem
import com.tapadoo.alerter.Alerter
import com.valdesekamdem.library.mdtoast.MDToast
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.net.NetworkInterface
import java.util.*

object Utility {
    fun getUniqueIMEIId(context: Context): String {
        try {
            val telephonyManager = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED) {
                return ""
            }
            var imei: String? = null
            imei = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                telephonyManager.imei
            } else {
                telephonyManager.deviceId
            }
            return if (imei != null && !imei.isEmpty()) {
                imei
            } else {
                Build.SERIAL
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return "not_found"
    }

    val macAddr: String
        get() {
            try {
                val all: List<NetworkInterface> = Collections.list(NetworkInterface.getNetworkInterfaces())
                for (nif in all) {
                    if (!nif.name.equals("wlan0", ignoreCase = true)) continue
                    val macBytes = nif.hardwareAddress ?: return ""
                    val res1 = StringBuilder()
                    for (b in macBytes) {
                        res1.append(String.format("%02X:", b))
                    }
                    if (res1.length > 0) {
                        res1.deleteCharAt(res1.length - 1)
                    }
                    return res1.toString().trim { it <= ' ' }.replace(":", "")
                }
            } catch (ex: Exception) {
            }
            return "02:00:00:00:00:00"
        }

    fun calSCode(context: Context?, numberPhone: List<String>): String {
            val n10 = numberPhone[numberPhone.lastIndex-1].toInt()
            val n9 = numberPhone[numberPhone.lastIndex - 2].toInt()
            val sCode = getItem(context, SaveItem.raw_Scode, "")
            if (sCode.length > 0) {
                val index = n9 + n10
                return sCode.substring(index, index + 10)
            }

        return ""
    }

    fun getNewVersion(context: Activity) {
        val retrofit = retrofitInstance
        val loginServer = retrofit!!.create(LoginServer::class.java)
        loginServer.getSecretCode(BuildConfig.VERSION_CODE.toString())!!.enqueue(object : Callback<SecretCodeModel?> {
            override fun onFailure(call: Call<SecretCodeModel?>, t: Throwable) {
                Log.e("fail get new version",t.message)
                MDToast.makeText(context,context.getString(R.string.error_in_connection),MDToast.LENGTH_LONG,MDToast.TYPE_ERROR).show()
            }

            override fun onResponse(call: Call<SecretCodeModel?>, response: Response<SecretCodeModel?>) {
                if (response.body()!!.result.equals("success", ignoreCase = true)) {
                    checkNewVersion(context, response.body())
                } else {
                    MDToast.makeText(context, response.body()!!.message, 2500, MDToast.TYPE_INFO).show()
                }
            }

        })
    }

    fun getSecretCode(context: Activity?) {
        val retrofit = retrofitInstance
        val loginServer = retrofit!!.create(LoginServer::class.java)
        loginServer.getSecretCode(BuildConfig.VERSION_CODE.toString())!!.enqueue(object : Callback<SecretCodeModel?> {
            override fun onFailure(call: Call<SecretCodeModel?>, t: Throwable) {
                TODO("Not yet implemented")
            }

            override fun onResponse(call: Call<SecretCodeModel?>, response: Response<SecretCodeModel?>) {
                if (response.body()!!.result.equals("success", ignoreCase = true)) {
                    val s_raw = response.body()!!.secretCode!!.trim { it <= ' ' }
                    setItem(context, SaveItem.raw_Scode, s_raw.substring(0, 30))
                } else {
                    MDToast.makeText(context, response.body()!!.message, 2500, MDToast.TYPE_INFO).show()
                }
            }

        })
    }

    private fun checkNewVersion(context: Activity, body: SecretCodeModel?) {
        if (body!!.versionCode == null) {
            return
        }
        if (body.versionCode!!.toInt() > BuildConfig.VERSION_CODE) {
            setItem(context, SaveItem.COME_NEW_VERSION, "1")
            setItem(context, SaveItem.NEW_VERSION_NAME, body.versionName)
            setItem(context, SaveItem.NEW_VERSION_URL, body.updateUrl)
        } else {
            setItem(context, SaveItem.COME_NEW_VERSION, "0")
        }
    }

    fun showAlertNewVersion(context: Activity, versionName: String, updateUrl: String?) {
        Alerter.create(context)
                .setTitle(context.getString(R.string.new_version))
                .setText(context.getString(R.string.version) + " " + versionName + " " + context.getString(R.string.available))
                .setDuration(10000)
                .setContentGravity(Gravity.CENTER)
                .setTitleTypeface(Typeface.createFromAsset(context.assets, "fonts/Vazir.ttf"))
                .setTextTypeface(Typeface.createFromAsset(context.assets, "fonts/sans.ttf"))
                .setIcon(R.drawable.ic_notifications_black_24dp)
                .setOnClickListener(object:View.OnClickListener{
                    override fun onClick(p0: View?) {
                        val uri = Uri.parse(updateUrl)
                        val goToMarket = Intent(Intent.ACTION_VIEW, uri)
                        goToMarket.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY or
                                Intent.FLAG_ACTIVITY_NEW_DOCUMENT or
                                Intent.FLAG_ACTIVITY_MULTIPLE_TASK)
                        try {
                            context.startActivity(goToMarket)
                        } catch (e: ActivityNotFoundException) {
                            context.startActivity(Intent(Intent.ACTION_VIEW,
                                    Uri.parse(updateUrl)))
                        }
                    }

                })
                .show()
    }

    fun calMID(numberPhone: List<String>): String {
        try {
            val n10 = numberPhone[numberPhone.lastIndex-1].toInt()
            val n9 = numberPhone[numberPhone.lastIndex-2].toInt()
            val mac = macAddr.split("")
            return "${n9}${mac[n9+1]}${n10}${mac[n10+1]}"
        } catch (e: Exception) {
            Log.e("eception:",e.message)
            return ""
        }
    }

    fun calApkId(context: Context?, numberPhone: List<String>): String {
        try {
            val mid = calMID(numberPhone)
            val uid = getItem(context, SaveItem.USER_ID, "")
            val len = uid.length
            val apkId ="${mid}${len}${uid}"
            setItem(context, SaveItem.APK_ID, apkId)
            return apkId
        } catch (e: Exception) {
            Log.e("Exception", e.message)
            return ""
        }
    }
}